debian-bo-master/bo/source/comm/xringd-1.10.orig/xringd.8
.\"
.\" Copyright (c) 1995-1996 Angelo Haritsis.
.\"
.\" xringd man page
.\"
.\" $Id$
.\"
.\" SH section heading
.\" SS subsection heading
.\" LP paragraph
.\" IP indented paragraph
.\" TP hanging label
.TH XRINGD 8
.SH NAME
xringd \- The Linux extended modem ring server
.SH SYNOPSIS
.B xringd
[
.I options
] 
[
.I modem_device_file
] 
.SH DESCRIPTION
.LP
The xringd Linux extended Ring server will listen on a modem device
for specific ring-delay patterns (sequences). Each sequence, when fully 
recognised, will execute a command you have chosen (subject to usual unix 
permissions). Delays are in fact delay ranges. Sequences and commands are 
read from a a configuration file.
.I xringd
does not disturb your other modem programs, not even your getty. It
coexists with them.
.I xringd
probes (asynchronously) for the actual RING signal on the serial line.

.SH OPTIONS
.TP
.B -a \fI command_on_each_ring
Run this command on every ring. Use perhaps to replace your boring phone
ring.
.TP
.B -c \fI config_file
Use an alternate configuration file. The default is
.B /etc/xringd.conf
.TP
.B -d 
Run in debug mode (no daemon - logging = 100).
.I xringd
does not run as a daemon and produces log messages on standard error.
.TP
.B -h | -?
See a mini usage info
.TP
.B -i msecs-ignored
If consecutive rings have a time (in msec) distance less than this one, they
are taken as one. For countries where a ring creates two sounds and
modems that subsequently cause two changes on the serial RI line. Use this
option to make two near RIs look as one to xringd. A value of 100-800 will
most likely be the most appropriate.
.TP
.B -l loglevel 
Logging level. Default=1. Use 10+ for more info. When running
as daemon you can use -l 10 or 100 to get debug messages via 
syslog(LOG_DEBUG, ...).  0 means NO logging at all.
.TP
.B -m \fI modem_device_file
The modem device file (can also be given as the final argument).
.TP
.B -n 
Performs only a syntax check of its configuration file. It implies
-d. Try this first when you write a new configuration file.
.I xringd
does not become a daemon and produces log messages on standard error.
.TP
.B -t \fI init_time
After a reset (or the first time it is run), the time (in seconds) to wait 
until rings are accepted. Default: 15

.SH CONFIGURATION FILE
The configuration file consists of lines of the following format:
.LP
.B R secs[-secs] [ R secs[-secs] ] ... : command
.LP
Each line is related to a sequence(pattern) that can be potentially
matched. The
.I command 
at the end gets executed if the sequence was fully matched. A full match
is found if the delays between the rings are within the delay ranges given
in the configuration line of a sequence. A full match will also reset 
the state machine. It will start accepting new rings as when run the 
first time.
.I R
means ring and it should 
.B always 
be the first symbol in a sequence.
.LP
Comment lines start with a ``#'' symbol at the beginning of the line. 
Empty lines are ignored.
.LP
Note that command lines options can also be included in the config file.
A line should start with the '-' of an option. See example below.
Options -c and -n are ignored in the config file.
Options in the configuration file take precedence over the
ones in the command line.

.SH EXAMPLE CONFIGURATION
.nf
# xringd configuration file -- sample
#
-a /usr/local/audio/bin/play /usr/local/lib/sounds/ring.au
-l 100
#
# 2 rings 10-16 sec apart followed by 30 secs silence
R 10-16 R 30 : /etc/ppp/ppp.start office1
# 3 rings 10-20 sec apart followed by 20 secs silence
R 10-20 R 10-20 R 20 : /etc/ppp/ppp.start office2
# 2 nearish rings then 1 ring after 20-26 secs, silence for 30 secs
R 1-5 R 20-26 R 30 : /usr/local/bin/turn-heater on

.SH FILES
.TP
.B /etc/xringd.conf
The default configuration file.
.TP
.B /dev/modem
The default modem device used.
.SH SIGNALS
The following signals have the specified effect when sent to the
.I xringd
process.
.TP
.B SIGINT, SIGTERM
Clean exit the server.
.TP
.B SIGUSR1
"Simulates" a RING as if it came from the modem.
.TP
.B SIGHUP
Restart the internal machine ignoring any current state. Reread
the config file. Close and reopen the modem device.  If a syntax error
is found in a line all the following lines are ignored. So when you
restart, make sure you look at the log for any reported errors. A better
way is to always "parse" your config file with "xringd -n" to check its
syntax first.

.SH NOTES
At the moment,
.I xringd
is device dependent to Linux kernel 1.3.48+ and serial devices that support
the TIOCMIWAIT, TIOCGICOUNT ioctl(2) calls. These were added by the same 
author to the Linux kernel so that a process can wait on a modem 
DCD,RI,DSR,CTS change on a serial port and can also read a kernel count of
the interrupts on each one of these 4 lines. I used the RI for this program.
(Other possibilities exist in using this ioctl for instrumentation projects.)
Note that only 16xx0 uarts implement this ioctl now (Jan96).
.LP
You
.B have
to use a proper serial cable for this to work. A cable with all pins
properly connected to your modem (especially the RI line for this program!)
and serial port will save you any trouble. Internal modems should normally 
work.
.LP
If you activate a program which uses the modem after ringd it should
normally
.B
flush
the input buffer. In many cases you will have a few
"RING" strings in your serial tty buffer that will most likely confuse
a dialup script (eg. chat).
.LP
The richness of the ring-delay pattern "language" is not great.
However, you certainly many possibilities.
Beware of overlaps though, and always have something
that will "unlock" any current sequence (eq. 4 consecutive rings that
safely exit from any current state).
.LP
If someone calls in while you are on the delay phase of your
"pattern" then you are obviously out of luck.
.LP
Only tone dialing phones allow quick dial that can meet short timing
restrictions possibly imposed by your configuration file. Make sure you
use the redial button on the calling phone if there is one - you will
be able to "dial" in about a second. Pulse dialers may introduce
unexpected delays. If they are your only choice, use longer delays and
wider delay ranges.
.LP
It was said to me that the "rings" you hear on the calling handset do not
directly correspond to the ones actually heard on the receiving end.
In the tests done with xringd in a few countries, the number of rings
remained the same on calling and called set. Just leave each one of the rings 
you hear on the calling end to "settle" (do not break them before they finish).
A delay between a ring heard on the caller set and the equivalent one on
the called one was noticed but causes no problem for xringd. Feel free
to send me your comments on this.
.LP
Many getty-like programs may be configured to pick up the phone on the
first ring. Obviously, this will make xringd minimally useful.
Make your getty to reply after 2-4 rings so that you have many possibilities
open for xringd.
.LP
pppd (and probably some other programs) like to hold a tty in exclusive mode.
Make sure you start xringd
.B
before
such programs, otherwise it won't be allowed to open the modem device.
Also, when such a program closes it may leave the line hung up. You need
to restart (kill -HUP) xringd in such a case.
It does not make sense to run xringd on a line which is permanently
used for PPP/SLIP - such a line never "rings"!
.LP
Spurious interrupts (and thus pseudo-RINGs) may occur during modem 
switch on/off; run xringd 
.B
after
your modem is switched on.
.LP
It is highly recommended - for security reasons -  to make the configuration 
file inaccessible (even for read) to anything but xringd. Treat it as
a shadow-password-like file. It is very easy for anyone to call your
number and activate a command, if they know a RING-delay sequence
"password". So try not to disarm your home-alarm via it.
You have been warned!

.SH AUTHOR
Angelo Haritsis (ah@doc.ic.ac.uk).


debian-bo-master/bo/source/comm/xringd-1.10.orig/xringd.c
/*
 * File:	xringd.c
 *
 * Synopsis:	Extensive modem ring daemon
 *
 * System:	Linux 1.3.48+
 *
 * Daemon that uses ringsm.c module to do interesting things with sequences
 * of modem RINGs. Does not disturb any getty running on the same
 * port as it only probes for the RI signal.
 * Device and OS dependent as it uses the TIOCMIWAIT/TIOCGICOUNT ioctls. The
 * ringsm can be used with "traditional" methods. This is just
 * cleaner as it does not disturb your system's getty.
 *
 * Configuration file default: /etc/xringd.conf
 * Config file lines are of this form:
 * R secs[-secs] [ R secs[-secs] ] ... : command
 * which specify the pattern of rings+delays leading to the exution of 
 * a command. 
 * 
 * Copyright (c) 1995-1996 Angelo Haritsis. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * $Id: xringd.c,v 1.5 1996/02/17 15:50:07 ah Exp ah $
 */

#ifndef linux
#error xringd only runs for linux - it uses 2 non-standard ioctls
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <paths.h>
#include <fcntl.h>
#include <stdarg.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <signal.h>
#include <termios.h>
#include <getopt.h>
#include <syslog.h>
#include <pwd.h>
#include <ctype.h>
#include <linux/version.h>
#include <linux/serial.h>

#include "ringsm.h"

#if !defined(LINUX_VERSION_CODE) || \
    (LINUX_VERSION_CODE < ((1 * 65536) + (3 * 256) + 48))
#error xringd needs Linux 1.3.48 or greater
#endif

#define RING_SIGNALS	/* define for raising SIG_RING on RING detection */
#define SIG_RING SIGUSR1
#define SIG_FLAGS 0	/* 0 or SA_RESTART (syscall after a sig) */

#define MAXLINE 512

#ifndef MODEM_FILE
#define MODEM_FILE "/dev/modem"
#endif
#ifndef CONF_FILE
#define CONF_FILE "/etc/xringd.conf"
#endif
#ifndef INIT_TIMEOUT
#define INIT_TIMEOUT 15
#endif

#define PROG "xringd"

#if 0
static char pidfilename[MAXPATHLEN] = "\0";
#endif

char *config_file = NULL;
char *modem_file = NULL;
char *do_onring = NULL;
int modem = -1; 
time_t reset_timeout = INIT_TIMEOUT;
int debug = 0;
int log_level = 1;
int syntax_only = 0;
long msec_ignore = 0;	/* 2 rings within this many msecs => taken as one */
sigset_t sig_mask;
int ringsm_parsed = FALSE;

void set_signals(void);
int modem_open(char *file, int reopen);
void usage_and_die(void);
int parse_an_arg(char opt, char *s, int in_conf_line);
int parse_an_arg_conf(char opt, char *s, int line);
void parse_args(int argc, char **argv);
char *str_new(char *s, char *new);
void show_args(void);
int is_ring_near(struct timeval *tm_bef);
void seq_success(char *seq_cmd);
void log_msg(int priority, int level, char *fmt, ...);
void cback_log_msg(int level, char *fmt, ...);
void daemon_init(void);
void ring_cmd(void);
void proc_ring(int i);
void sig_child(int);
void sig_hup(int i);
void sig_alarm(int i);
void clean_exit(int code);
void sig_bye(int i);
void fork_cmd(char *cmd);


void
main (int argc, char *argv[])
{
	int arg;
	struct serial_icounter_struct c;
	struct timeval tm_bef;
	static char str_near[100];

	config_file = str_new(config_file, CONF_FILE);	/* initialise strings */
	modem_file = str_new(modem_file, MODEM_FILE);
	parse_args(argc, argv);
	if (access(config_file, R_OK) != 0) {
		log_msg(-1, 0, "cannot open config file %s", config_file);
		usage_and_die();
		exit(1);
	}
	if ((arg = ringsm_parse(config_file, &parse_an_arg_conf)) < 0) {
		log_msg(-1, 0, "error in config file, line %d", -arg);
		exit(1);
	}
	ringsm_parsed = TRUE;
	if (syntax_only) {
		log_msg(-1, 0, "config file %s -> syntax ok", config_file);
		exit(0);
	}
	modem = modem_open(modem_file, FALSE);
	if (ioctl(modem, TIOCGICOUNT, &c) != 0) {
		log_msg(-1, 0, "needs Linux 1.3.48+, 16xx0 uart");
		usage_and_die();
		exit(1);
	}
	show_args();

	/* Daemon proper */
	daemon_init();
	if (!debug && log_level > 0)
		openlog(PROG, LOG_PID, LOG_DAEMON);
	set_signals();
	ringsm_init(reset_timeout, &seq_success, &alarm, &cback_log_msg);
	str_near[0] = 0;
	if (msec_ignore > 0)
		sprintf(str_near, "; rings %ld msec or less apart ignored", msec_ignore);
	log_msg(LOG_INFO, 1, "version %s started by %s (uid %d)%s", VERSION, getpwuid(getuid())->pw_name, getuid(), str_near);

	gettimeofday(&tm_bef, NULL);
	/* main daemon loop */
	while (1) {
		/* wait on RI line */
		if (ioctl(modem, TIOCMIWAIT, TIOCM_RNG) != 0) {
			if (EINTR == errno) {
				log_msg(LOG_DEBUG, 1000, "ioctl restarting - signal");
				continue;
			}
			log_msg(LOG_ERR, 10, "%%m - reopening %s", modem_file);
			/*
			 * XXX EIO may occur: port hung up by other process!
			 * Reopen it - danger: reopening failing continously
			 */
			close(modem);
			modem = modem_open(modem_file, TRUE);
			continue;
		}
		/* we got a ring; if near the previous, ignore it */
		if (is_ring_near(&tm_bef))
			continue;
		/* ring is certain; notify the state machine */
#ifdef RING_SIGNALS
		raise(SIG_RING);
#else
		proc_ring(arg /*dummy*/);
#endif
		if (debug)
			ringsm_dump();
	}
	/* NOTREACHED */
}

void
show_args(void)
{
#ifdef DEBUG
	log_msg(-1, 100, "Modem: %s - Config: %s\nOnring: %s\n"
			"reset: %d - ignore: %d - debug: %d - log: %d\n",
			modem_file, config_file, do_onring,
			reset_timeout, msec_ignore, debug, log_level);
#endif
}

void
usage_and_die(void)
{
	fprintf(stderr, "xringd version " VERSION " - by A. Haritsis (ah@doc.ic.ac.uk)\n"
			"usage: %s [-a rngcmd] [-c cfgfile] [-d] [-h] [-i ignore_msec] [-l loglevel]\n"
			"              [-m modem_dev] [-n] [-t initime] [modem_dev]\n", PROG);
	exit(1);
}

void
set_signals(void)
{
#define SIGNAL(s, flags, handler, old) do { \
	sa.sa_handler = handler; \
	sa.sa_flags = flags; \
	if (sigaction(s, &sa, old) < 0) { \
		log_msg(LOG_ERR, 1, "error setting signal (%%m)"); \
		clean_exit(1); \
	} \
} while (0)

	struct sigaction sa;   

	sigemptyset(&sig_mask);
	sigaddset(&sig_mask, SIGALRM);	/* only one signal served at a time */
	sigaddset(&sig_mask, SIGCHLD);
	sigaddset(&sig_mask, SIGHUP);
	sigaddset(&sig_mask, SIGINT);
	sigaddset(&sig_mask, SIGTERM);
	sigaddset(&sig_mask, SIGQUIT);
#ifdef RING_SIGNALS
	sigaddset(&sig_mask, SIG_RING);
#endif
	sa.sa_mask = sig_mask;
	SIGNAL(SIGALRM, SIG_FLAGS, sig_alarm, NULL);
	SIGNAL(SIGCHLD, SIG_FLAGS, sig_child, NULL);
#ifdef RING_SIGNALS
	SIGNAL(SIG_RING, SIG_FLAGS, proc_ring, NULL);
#endif
	SIGNAL(SIGHUP, SIG_FLAGS, sig_hup, NULL);
	SIGNAL(SIGINT, SIG_FLAGS, sig_bye, NULL);
	SIGNAL(SIGTERM, SIG_FLAGS, sig_bye, NULL);
	SIGNAL(SIGQUIT, SIG_FLAGS, sig_bye, NULL);
#undef SIGNAL
}

void
log_msg(int priority, int level, char *fmt, ...)
{
	va_list p;
	char buf[MAXLINE];

	va_start(p, fmt);
	vsprintf(buf, fmt, p);
	if (log_level >= level)
		if (debug || -1 == priority) {
			fflush(stdout);
			fprintf(stderr, PROG ": %s\n", buf);
			fflush(stderr);
		} else
			syslog(priority, buf);
	va_end(p);
}


void
cback_log_msg(int level, char *fmt, ...)
{
	va_list p;
	char buf[MAXLINE];

	va_start(p, fmt);
	vsprintf(buf, fmt, p);
	log_msg(LOG_DEBUG, level, buf);
	va_end(p);
}

/*
 * is_ring_near -- is ring we just got nearer than msec_ignore msec to 
 *		   the previous?
 */
int
is_ring_near(struct timeval *tm_bef)
{
	struct timeval tm;
	long msec;
	int ret;

	if (0 == msec_ignore)
		return FALSE;

	gettimeofday(&tm, NULL);
	msec = (tm.tv_sec - tm_bef->tv_sec) * 1000L + 
	       (tm.tv_usec - tm_bef->tv_usec) / 1000L;
	*tm_bef = tm;	/* update the before time! */
	ret = (msec <= msec_ignore);
	if (ret)
		log_msg(LOG_DEBUG, 10, "RI near (%ld msec) - dumped", msec);
	return ret;
}

/*
 * called via ringsm_process_timeout
 */
void
seq_success(char *seq_cmd)
{
	log_msg(LOG_INFO, 1, "launching: %s", seq_cmd);
	fork_cmd(seq_cmd);
}

/*
 * Called when a RING signal appears
 */
void
proc_ring(int i)
{
	static int rings = 1;
	int ret;

	if (do_onring)
		ring_cmd();
	log_msg(LOG_DEBUG, 10, "RING #%d", rings++);
	ret = ringsm_process_a_ring();
	if (!ret)
		log_msg(LOG_DEBUG, 100, "ignored ring #%d",  rings-1);
}

void
sig_child(int i)
{
	int status;

	/* we should probably check the full exit status of the dead child */
	wait(&status);
}

void
sig_hup(int i)
{
	int arg;

	if (!ringsm_parsed)
		return;
	log_msg(LOG_INFO, 1, "HUP received: resetting");
	ringsm_reset(FALSE);
	/* close and reread config file */
	ringsm_close();
	if ((arg = ringsm_parse(config_file, &parse_an_arg_conf)) < 0)
		log_msg(LOG_ERR, 0, "error in config file, line %d - ignoring the next lines", -arg);
	if (modem >= 0) {
		close(modem);
		modem = modem_open(modem_file, TRUE);
	}
	ringsm_reset(TRUE);
	show_args();
}

void
sig_alarm(int i)
{
	ringsm_process_timeout();
}

void
clean_exit(int code)
{
	log_msg(LOG_INFO, 1, "Exit");
	/*if (unlink(pidfilename) < 0 && errno != ENOENT)
        	log_msg(LOG_WARNING, 1, "unable to delete pid file: %m");
        pidfilename[0] = 0;*/
	closelog();
	exit(code);
}

void
sig_bye(int i)
{
	ringsm_reset(FALSE);
	clean_exit(0);
}

#if 0
void
setdtr(int fd, int on)
{
	int modembits = TIOCM_DTR;

	ioctl(fd, (on ? TIOCMBIS : TIOCMBIC), &modembits);
}
#endif

int
modem_open(char *file, int reopen)
{
	int fd = -1;

	if ((fd = open(file, O_RDONLY | O_NOCTTY | O_NDELAY)) < 0) {
		if (reopen) {
			log_msg(LOG_ERR, 1, "error reopening %s (%%m)", file);
			usleep(3*1000000);	/* uses select - not SIGALRM */
			return fd;
		} else {
			log_msg(-1, 0, "error opening modem device");
			perror(NULL);
			exit(1); 
		}
	}
	return fd;
}

#if 0
void
write_pid(void)
{
	FILE *pidf;

        (void) sprintf(pidfilename, "%s%s.pid", _PATH_VARRUN, PROG);
        if ((pidf = fopen(pidfilename, "w")) != NULL) {
            fprintf(pidf, "%d\n", getpid());
            (void) fclose(pidf);
        } else {
            log_msg(LOG_ERR, 1, "Failed to create pid file %s: %m", pidfilename);
            pidfilename[0] = 0;
        } 
}
#endif

void
daemon_init(void)
{
	int pid;

	if (debug)
		return;
	if ( (pid = fork()) < 0) {
		perror(NULL);
		exit(1);
	} else if (pid != 0)
		exit(0);	/* parent goes */
	/* child */
	setsid();
	chdir("/");
	umask(S_IRWXG | S_IRWXO);
	close(STDIN_FILENO);
	close(STDOUT_FILENO);
	close(STDERR_FILENO);
}

void
ring_cmd(void)
{
	char *p;

	for (p = do_onring; *p != 0; p++)
		if (!isspace(*p))
			break;
	if (*p == 0)		/* all whitespace */
		return;
	log_msg(LOG_INFO, 10, "running (per ring): %s", do_onring);
	fork_cmd(do_onring);
}

/*
 * fork_cmd --- fork a shell to execute command cmd 
 *
 * Establish a "clean slate" first.
 */
void
fork_cmd(char *cmd)
{
	int pid;
	int fdnull;

	if ((pid = fork()) < 0) {
		log_msg(LOG_ERR, 1, "error forking for %s (%%m)", cmd);
		return;
	} else if (0 == pid) {
		/* child */
		/* unblock signals we use - may be blocked here */
		sigprocmask(SIG_UNBLOCK, &sig_mask, NULL);
		setsid();
		close(STDIN_FILENO);
		close(STDOUT_FILENO);
		close(STDERR_FILENO);
		if (modem >= 0)
			close(modem);
		/* connect /dev/null to std{in/out/err} */
		fdnull = open(_PATH_DEVNULL, O_RDWR);
		if (fdnull >= 0) {
			if (fdnull != 0) {
				dup2(fdnull, 0);
				close(fdnull);
			}
			dup2(0, 1);
			dup2(0, 2);
		}
		setuid(getuid());
		setgid(getgid());
		execl("/bin/sh", "sh", "-c", cmd, NULL);
		/* could not run cmd ; report it */
		log_msg(LOG_ERR, 1, "error running: /bin/sh -c %s (%%m)", cmd);
		exit(1);
	}
	/* parent */
}

int
parse_an_arg(char opt, char *s, int conf_line)
{
	switch (opt) {
	case 'a':
		do_onring = str_new(do_onring, s);
		break;
	case 'c':
		if (0 == conf_line)
			config_file = str_new(config_file, s);
		break;
	case 'd':
		debug++;
		log_level = 100;
		break;
	case 'i':
		msec_ignore = atoi(s);
		break;
	case 'l':
		log_level = atoi(s);
		break;
	case 'm':
		modem_file = str_new(modem_file, s);
		break;
	case 'n':
		if (0 == conf_line) {
			syntax_only++;
			debug++;
		}
		break;
	case 't':
		reset_timeout = atoi(s);
		break;

	case '?':
	case 'h':
		if (0 == conf_line)
			usage_and_die();
		break;
	default:
		if (0 == conf_line)
			usage_and_die();
		else
			log_msg(LOG_ERR, 0, "option %c error in conf file, line %d - ignoring", opt, conf_line);
		break;
	}
	return TRUE;
}

/* 
 * parse_an_arg_conf -- callback for an argument parse (from config file)
 */
int
parse_an_arg_conf(char opt, char *s, int line)
{
	return parse_an_arg(opt, s, line);
}

void
parse_args(int argc, char **argv)
{
	int opt;

	opterr = 0;
	while ( (opt = getopt(argc, argv, "a:c:dl:i:l:m:nt:h?")) != EOF )
		parse_an_arg((char) opt, optarg, 0);
	if (optind < argc)
		modem_file = str_new(modem_file, argv[optind]);
}

/* 
 * str_new -- free old space if alloc'ed; alloc new and assign new value
 * XXX new is NULL or "" !!
 */
char *
str_new(char *s, char *new)
{
	int newsize = strlen(new) + 1;
	char *p;

	if (NULL == s)
		p = malloc(newsize);
	else
		p = realloc(s, newsize);
	strcpy(p, new);
	return p;
}
